Debug-frontend-tests.md

```diff
@@ -0,0 +1,72 @@
+	    at ZoneDelegate../node_modules/zone.js/dist/zone.js.ZoneDelegate.invoke (core/templates/pages/about-page/about-page.module.js:237518:36)
+	    at Zone../node_modules/zone.js/dist/zone.js.Zone.run (core/templates/pages/about-page/about-page.module.js:237276:47)
+	    at runInTestZone (core/templates/combined-tests.spec.js:427336:38)
+	    at UserContext.<anonymous> (core/templates/combined-tests.spec.js:427351:24)
+Chrome Headless 83.0.4103.116 (Mac OS 10.15.5): Executed 1 of 2277 (1 FAILED) (0 secs / 0.046 secs)
+```
+
+The `zone.js` stack elements relate to the [zone.js](https://github.com/angular/angular/tree/master/packages/zone.js) package that is handling our asynchronous operations. Notice that the line numbers are unreasonably large. This is because to run our tests, we load all our testing code and dependencies into `combined-tests.spec.js`. The line numbers refer to the resulting "mega-file," which doesn't actually exist on the file system.
+
+Below, we'll discuss some strategies you can use to overcome these difficulties.
+
+## Find tests
+
+You can use test names to find test code files, but you can't just search for the whole name. You have to understand how test names are constructed. For example, here's the kind of test structure that produced the test name "Assets Backend API Service on dev mode should handle rejection when fetching a file fail" above:
+
+```js
+describe('Assets Backend API Service', () => {
+  describe('on dev mode', () => }
+    it('should handle rejection when fetching a file fails', () => {
+      ...
+    });
+  });
+});
+```
+
+The strings from the `describe` and `it` functions get concatenated with space (` `) delimiters to create the test name. The name of the `it` block usually starts with `should`, so we recommend searching the codebase for the portion of the test name from `should` to the end (e.g. "should handle rejection when fetching a file fails" in this case).
+
+## Isolate tests
+
+Once you have identified which test is failing, you often want to run that test in isolation next. This will let you get results quickly as you try other debugging strategies. Thankfully, you just have to change the `it()` of your test to `fit()`. You can even do this multiple times to mark multiple tests for running. Then when you run the frontend tests, only those tests with `fit()` will run. Note that `fdescribe()` works the same way.
+
+**Remember to revert your `fit()` and `fdescribe()` changes before committing!**
+
+## Verbose mode
+
+By default, we suppress log output from the tests to the terminal when running the frontend tests. This keeps the test output clean so it's easy to find which tests failed. However, this output can be useful for debugging. For example, if you add `console.log()` or `console.error()` statements to your tests for debugging, the output from those statements will be suppressed by default. To view this output, pass `--verbose` when you run the frontend tests:
```

Comment:
This is just for debugging, so these statements shouldn't be committed. In that case, the linter doesn't matter, right?

Reply:
During our, GSoC to debug or test the flake fix we used the following [code](https://gist.github.com/gp201/3bb7e649680b01eed8d3370c7e44e506). We then pushed this to our fork and created a PR to our fork so the GitHub actions would run on our fork.

Reply:
Okay. I'll add a note about the linter. I think `console.log` is actually useful because if you aren't testing on CI, you might _want_ the linter to remind you in case you forget to remove your debugging statements

Reply:
done